rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +86,107 @@ struct @(spec.base_type.type)_
-    elif field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, cpp_value), end='')
-        leading_colon = ' '
-if leading_colon == ' ':
-    print()
+# The creation of the constructors for messages is a bit complicated.  The goal
+# is to have a constructor where the user can control how the fields of the
+# message get initialized via the _init parameter to the constructor.  See
+# http://design.ros2.org/articles/generated_interfaces_cpp.html#constructors
+# for a detailed explanation of the different _init parameters.
+init_list, alloc_list, member_list = create_init_alloc_and_member_lists(spec)
+
+def generate_default_string(membset):
+    strlist = []
+    for memb in membset.members:
+        if memb.default_value is not None:
+            if isinstance(memb.default_value, list):
+                if not memb.default_value or memb.default_value.count(memb.default_value[0]) == len(memb.default_value):
+                    strlist.append("std::fill(this->%s.begin(), this->%s.end(), %s);" % (memb.name, memb.name, memb.default_value[0]))
+                else:
+                    for index, val in enumerate(memb.default_value):
+                        strlist.append("this->%s[%d] = %s;" % (memb.name, index, val))
+            else:
+                strlist.append("this->%s = %s;" % (memb.name, memb.default_value))
+
+    return strlist
+
+def generate_zero_string(membset, fill_args):
+    strlist = []
+    for memb in membset.members:
+        if isinstance(memb.zero_value, list):
+            if memb.zero_need_array_override:
+                strlist.append("this->%s.fill(%s {%s});" % (memb.name, memb.type, fill_args))
+            else:
+                strlist.append("std::fill(this->%s.begin(), this->%s.end(), %s);" % (memb.name, memb.name, memb.zero_value[0]))
+        else:
+            strlist.append("this->%s = %s;" % (memb.name, memb.zero_value))
+    return strlist
 }@
-@[  if alloc_name == '_alloc']@
-// *INDENT-ON*
-@[  end if]@
+  explicit @(spec.base_type.type)_(rosidl_generator_cpp::MessageInitialization _init = rosidl_generator_cpp::MessageInitialization::ALL)
+@[if init_list]@
+  : @(',\n    '.join(init_list))
+@[end if]@
   {
-@# generate constructor body
-@[  if alloc_name and not used_alloc_name]@
-    (void)@(alloc_name);
-@[  end if]@
-@{
-for field in spec.fields:
-    # default values of dynamic arrays with allocators if available
-    if alloc_name and field.type.is_array and \
-            (field.type.array_size is None or field.type.is_upper_bound) and \
-            field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        print('    %s = %s;' % (field.name, cpp_value))
-}@
+@[if len(member_list) == 0]@
```

Comment:
Simpler: `if not member_list:`

Same below.

Reply:
Done in 114078e